worlds/saving_princess/__init__.py

```diff
@@ -0,0 +1,156 @@
+                t = self.music_table[song_id]
+                self.music_table[song_id] = song_id
+                self.music_table[song_index] = t
+
+    def create_regions(self) -> None:
+        from .Regions import create_regions
+        create_regions(self.multiworld, self.player, self.is_pool_expanded)
+
+    def create_items(self) -> None:
+        items_made: int = 0
+
+        # now, for each item
+        item_dict = Items.item_dict_expanded if self.is_pool_expanded else Items.item_dict_base
+        for item_name, item_data in item_dict.items():
+            # create count copies of the item
+            for i in range(item_data.count):
+                self.multiworld.itempool.extend([self.create_item(item_name)])
+            items_made += item_data.count
+            # and create count_extra useful copies of the item
+            original_item_class: ItemClass = item_data.item_class
+            item_data.item_class = ItemClass.useful
+            for i in range(item_data.count_extra):
+                self.multiworld.itempool.extend([self.create_item(item_name)])
+            item_data.item_class = original_item_class
+            items_made += item_data.count_extra
+
+        # get the number of unfilled locations, that is, locations for items - items generated
+        location_count = len(Locations.location_dict_base)
+        if self.is_pool_expanded:
+            location_count = len(Locations.location_dict_expanded)
+        junk_count: int = location_count - items_made
+
+        # and generate as many junk items as unfilled locations
+        for i in range(junk_count):
+            self.multiworld.itempool.extend([self.create_item(self.get_filler_item_name())])
+
+    def create_item(self, name: str) -> Items.SavingPrincessItem:
+        return Items.item_dict[name].create_item(self.player)
+
+    def get_filler_item_name(self) -> str:
+        filler_list = list(Items.item_dict_filler.keys())
+        # check if this is going to be a trap
+        if self.random.randint(0, 99) < self.options.trap_chance:
+            filler_list = list(Items.item_dict_traps.keys())
+        # and return one of the names at random
+        return self.random.choice(filler_list)
+
+    def fill_slot_data(self) -> Dict[str, Any]:
+        return {
+            "death_link": self.options.death_link.value,
+            "expanded_pool": self.options.expanded_pool.value,
+            "instant_saving": self.options.instant_saving.value,
+            "sprint_availability": self.options.sprint_availability.value,
+            "cliff_weapon_upgrade": self.options.cliff_weapon_upgrade.value,
+            "ace_weapon_upgrade": self.options.ace_weapon_upgrade.value,
+            "shake_intensity": self.options.shake_intensity.value,
+            "iframes_duration": self.options.iframes_duration.value,
+            "music_table": self.music_table,
+        }
+
+    def set_rules(self):
+        from .Rules import set_rules
+        set_rules(self)
```

Comment:
This should probably be moved to line up with when it occurs in the call order, namely after `create_items` / before `fill_slot_data`